pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -1107,106 +1292,95 @@ func (configgen *ConfigGeneratorImpl) buildSidecarOutboundListenerForPortOrUDS(l
-				// So append the new filter chains to the existing listener's filter chains
-				newFilterChains = append(newFilterChains, incomingFilterChain)
-				if pluginParams.Service != nil {
-					lEntry := listenerMap[listenerMapKey]
-					lEntry.services = append(lEntry.services, pluginParams.Service)
-				}
+	// There are 9 types conflicts
+	//    Incoming Existing
+	//  1. HTTP -> HTTP
+	//  2. HTTP -> TCP
+	//  3. HTTP -> unknown
+	//  4. TCP  -> HTTP
+	//  5. TCP  -> TCP
+	//  6. TCP  -> unknown
+	//  7. unknown -> HTTP
+	//  8. unknown -> TCP
+	//  9. unknown -> unknown
+	//  Type 1 can be resolved by appending service to existing services
+	//  Type 2 can be resolved by merging TCP filter chain with HTTP filter chain
+	//  Type 3 can be resolved by appending service to existing services
+	//  Type 4 can be resolved by merging HTTP filter chain with TCP filter chain
+	//  Type 5 can be resolved by merging TCP filter chains
+	//  Type 6 can be resolved by merging TCP filter chains
+	//  Type 7 can be resolved by appending service to existing services
+	//  Type 8 can be resolved by merging TCP filter chains
+	//  Type 9 can be resolved by merging TCP and HTTP filter chains
+
+	switch conflictType {
+	case NoConflict:
+		if currentListenerEntry != nil {
+			currentListenerEntry.listener.FilterChains = mergeTCPFilterChains(mutable.Listener.FilterChains,
+				pluginParams, listenerMapKey, listenerMap, node)
+		} else {
+			listenerMap[listenerMapKey] = &outboundListenerEntry{
+				services:    []*model.Service{pluginParams.Service},
+				servicePort: pluginParams.Port,
+				bind:        listenerOpts.bind,
+				listener:    mutable.Listener,
+				protocol:    pluginParams.Port.Protocol,
 			}
 		}
-		currentListenerEntry.listener.FilterChains = newFilterChains
-	} else {
+
+	case HTTPOverHTTP, HTTPOverAuto:
+		// Append service only
+		currentListenerEntry.services = append(currentListenerEntry.services, pluginParams.Service)
+	case HTTPOverTCP:
+		// Merge HTTP filter chain to TCP filter chain
+		currentListenerEntry.listener.FilterChains = mergeFilterChains(mutable.Listener.FilterChains, currentListenerEntry.listener.FilterChains)
+		currentListenerEntry.protocol = protocol.Unsupported
+		currentListenerEntry.listener.ListenerFilters =
+			append(currentListenerEntry.listener.ListenerFilters, &listener.ListenerFilter{Name: envoyListenerHTTPInspector})
+		currentListenerEntry.services = append(currentListenerEntry.services, pluginParams.Service)
+
+	case TCPOverHTTP:
+		// Merge TCP filter chain to HTTP filter chain
+		currentListenerEntry.listener.FilterChains = mergeFilterChains(currentListenerEntry.listener.FilterChains, mutable.Listener.FilterChains)
+		currentListenerEntry.protocol = protocol.Unsupported
+		currentListenerEntry.listener.ListenerFilters =
+			append(currentListenerEntry.listener.ListenerFilters, &listener.ListenerFilter{Name: envoyListenerHTTPInspector})
```

Comment:
you cannot do this if the tcp is mysql/mongo. this is why the well known port checks needs to be here and not in model

Reply:
I think well known port checks should be both places.